src/Disks/ObjectStorages/VFSTransactionLog.cpp

```diff
@@ -0,0 +1,133 @@
+#include "VFSTransactionLog.h"
+#include "Common/ZooKeeper/ZooKeeper.h"
+#include "IO/ReadHelpers.h"
+#include "VFSTraits.h"
+#include "base/EnumReflection.h"
+#include "base/defines.h"
+#include "base/find_symbols.h"
+
+namespace DB
+{
+namespace ErrorCodes
+{
+extern const int LOGICAL_ERROR;
+}
+
+String VFSTransactionLogItem::serialize() const
+{
+    if (type == Type::CreateInode)
+        return fmt::format("{} {} {} {}", type, remote_path, local_path, bytes_size);
+    return fmt::format("{} {}", type, remote_path);
+}
+
+VFSTransactionLogItem VFSTransactionLogItem::deserialize(std::string_view str)
+{
+    VFSTransactionLogItem out;
+
+    // TODO myrrc proper checking
+    std::vector<std::string_view> parts;
+    splitInto<' '>(parts, str);
+
+    if (parts.size() == 4)
+    {
+        out.type = Type::CreateInode;
+        out.remote_path = parts[1];
+        out.local_path = parts[2];
```

Comment:
single local path is not enough (see below), - same inode can have many local_paths (also from different replicas)

Reply:
The general issue is that local_path is propagated only in e.g. `writeBuffer` so it's something like `tmp_merge_ololo_0_0`. When the real part is renamed, we don't push renaming updates into Keeper. Personally I believe that for this task we don't need local_path, as we use it only to check snapshot entries in log.